pkg/controller/addon/klusterlet_addon_controller.go

```diff
@@ -0,0 +1,543 @@
+				return false
+			}
+			return agentv1.KlusterletAddons[e.MetaNew.GetName()]
+		},
+	})
+}
+
+type ReconcileKlusterletAddOn struct {
+	client client.Client
+}
+
+func (r *ReconcileKlusterletAddOn) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the managedCluster instance
+	managedCluster := &managedclusterv1.ManagedCluster{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: request.Namespace}, managedCluster); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Fetch the klusterletAddonConfig instance
+	klusterletAddonConfig := &agentv1.KlusterletAddonConfig{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, klusterletAddonConfig); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, r.deleteAllManagedClusterAddon(managedCluster.GetName())
+		}
+		return reconcile.Result{}, err
+	}
+
+	if isPaused(klusterletAddonConfig) {
+		return reconcile.Result{}, nil
+	}
+
+	nodeSelector, err := getNodeSelector(managedCluster)
+	if err != nil {
+		return reconcile.Result{}, nil
+	}
+
+	var aggregatedErrs []error
+	for addonName, enable := range agentv1.KlusterletAddons {
+		if !enable {
+			continue
+		}
+		if !addonIsEnabled(addonName, klusterletAddonConfig) {
+			if err := r.deleteManagedClusterAddon(addonName, managedCluster.GetName()); err != nil {
+				aggregatedErrs = append(aggregatedErrs, err)
+			}
+			continue
+		}
+
+		imageOverrides, err := getImageOverrides(r.client, managedCluster, addonName)
+		if err != nil {
+			return reconcile.Result{}, nil
+		}
+		gv := getGlobalValues(nodeSelector, imageOverrides, addonName, klusterletAddonConfig)
+
+		if err := r.updateManagedClusterAddon(gv, addonName, managedCluster.GetName()); err != nil {
+			aggregatedErrs = append(aggregatedErrs, err)
+		}
+	}
+	if len(aggregatedErrs) != 0 {
+		return reconcile.Result{}, fmt.Errorf("failed create/update addon %v", aggregatedErrs)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileKlusterletAddOn) deleteAllManagedClusterAddon(clusterName string) error {
+	var aggregatedErrs []error
+	for addonName, enable := range agentv1.KlusterletAddons {
+		if !enable {
+			continue
+		}
+		if addonName == agentv1.WorkManagerAddonName {
+			continue
+		}
+		err := r.deleteManagedClusterAddon(addonName, clusterName)
+		if err != nil {
+			aggregatedErrs = append(aggregatedErrs, err)
+		}
+	}
+	if len(aggregatedErrs) != 0 {
+		return fmt.Errorf("failed to delelte all addons %v", aggregatedErrs)
+	}
+	return nil
+}
+
+func (r *ReconcileKlusterletAddOn) deleteManagedClusterAddon(addonName, clusterName string) error {
+	if !agentv1.KlusterletAddons[addonName] {
+		return nil
+	}
+	if addonName == agentv1.WorkManagerAddonName {
+		return nil
+	}
+	addon := &addonv1alpha1.ManagedClusterAddOn{}
+	err := r.client.Get(context.TODO(), types.NamespacedName{Name: addonName, Namespace: clusterName}, addon)
```

Comment:
I think you don't have to `get` the addon before deleting it.
https://github.com/kubernetes-sigs/controller-runtime/blob/273e608d8a6cd7f34b6b56a6bd4626878319aed7/pkg/client/example_test.go#L219

Reply:
done